synapse/storage/state.py

```diff
@@ -681,6 +702,70 @@ def _get_state_for_groups(self, groups, types=None, filtered_types=None):
                 list of event types.  Other types of events are returned unfiltered.
                 If None, `types` filtering is applied to all events.
 
+        Returns:
+            Deferred[dict[int, dict[(type, state_key), EventBase]]]
+                a dictionary mapping from state group to state dictionary.
+        """
+        if types is not None:
```

Comment:
I'm finding the multiple calls to `_get_state_for_groups_using_cache` quite hard to reason about, and the special-case of checking for `filtered_types == [EventTypes.Member]` is a bit sad (apart from the case of longer lists including `EventTypes.member`, what if filtered_types is actually a tuple or a set? I know the docstring claims it should be a list but I can't see any reason for that, and the failure mode will be subtle if someone gets it wrong).

How about:

```python
        if types is not None:
            non_member_types = [t for t in types if t[0] != EventTypes.Member]

            if filtered_types is not None and EventTypes.Member not in filtered_types:
                # we want all of the membership events
                member_types = None
            else:
                member_types = [t for t in types if t[0] == EventTypes.Member]

        else:
            non_member_types = None
            member_types = None

        non_member_state = yield self._get_state_for_groups_using_cache(
            groups, self._state_group_cache, non_member_types, filtered_types,
        )
        member_state = yield self._get_state_for_groups_using_cache(
            groups, self._state_group_members_cache, member_types, None,
        )
```

Reply:
This is certainly much terser, but ironically i personally find it much more cryptic and hard to reason about, whereas the simple symmetricity of "if we're lazy-loading members, split the query intelligently. if we're filtering on types, split the query naively. otherwise, split the query without filtering" felt clearer by spelling out the flows we care about and following the same pattern for each branch.

However, I don't have strong feelings, and it's nice that this handles the whole "what if filtered_types is a longer list that contains EventTypes.Member" scenario better (which I'd considered, but fell through to the naive handler).  So i've gone with it (with a comment to explain the somewhat magical 'None' in the final line).

Reply:
I think if the param was "we're lazy-loading members", I'd agree with you. But for better or worse, it's not.